Add basic control system protocols and measuring

wthrowe opened the pull request:
Dependent on #3467, which is included here. Only the last commit is new.

This only includes the measurement-related part of the protocols.  More things will be added later as necessary.

Apologies for the single large commit, but most of this has to be tested as a single unit, and I didn't want to add code independent of its tests.

There is one doxygen error that I have not figured out how to quiet.

## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Add basic control system protocols and measuring

The documentation refers to a control_system::Trigger class that will
be added in a future commit.

wthrowe commented:
This will probably need some updates once #3479 is done, so marking in-progress.

wthrowe commented:
Significant changes after incorporating #3479, but no longer dependent.

knelli2 commented:
I haven't looked at this too much yet, but it looks like this has the events for measuring, but no triggers. Will those be added in a separate PR?

wthrowe commented:
Yes.  I didn't really want to include the event here either, but I decided it was the only reasonable way to check that the examples for the documentation were valid.

wthrowe commented:
Added two trivial "NamespaceDocs.hpp" files for `control_system` and `control_system::protocols` to make them show up in doxygen.

wthrowe commented:
Rebased and adjusted the wording on the submeasurement interpolation_target_tag.  (There is only one per submeasurement, not one per control system as implied before.)

wthrowe commented on tests/Unit/Helpers/ControlSystem/Examples.hpp:
```diff
@@ -0,0 +1,137 @@
+  using type = double;
+};
+
+struct SomeControlSystemUpdateFunctor {
```
Change self-request: This is not a functor, and also does not need constexpr if guards.  Also check if any documentation refers to it as a functor.

knelli2 commented on src/ControlSystem/Event.hpp:
```diff
@@ -0,0 +1,125 @@
+      tmpl::bind<process_measurement_for_control_system, tmpl::_1>>;
+
+ public:
+  /// \cond
+  explicit Event(CkMigrateMessage* /*unused*/) noexcept {}
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(Event);  // NOLINT
+  /// \endcond
```
Probably also wrap these in LCOV, unless you can serialize and deserialize in a test? If that actually makes sense to do?

wthrowe replied:
I think these are for checkpointing or something.  Not sure.

knelli2 commented on src/ControlSystem/Event.hpp:
```diff
@@ -0,0 +1,125 @@
+    tmpl::for_each<typename measurement::submeasurements>(
+        [&array_index, &box, &cache, &component,
+         &measurement_id](auto submeasurement) noexcept {
+          using Submeasurement = tmpl::type_from<decltype(submeasurement)>;
+          db::apply<Submeasurement>(box, measurement_id, cache, array_index,
+                                    component, ControlSystems{});
```
Overall, I was having a hard time understanding what components everything was run on. Adding a comment either before the operator or before the `db::apply` saying that these events are run on the elements would go a long way. Similarly with the other `apply`s and such in this PR.

wthrowe replied:
I'll add it to the protocol documentation.  That is probably more visible.

knelli2 commented on src/ControlSystem/Event.hpp:
```diff
@@ -0,0 +1,125 @@
+}  // namespace evolution::Tags
+/// \endcond
+
+namespace control_system {
+// This Event is tested in Test_Measurement.cpp
+
+/// \ingroup ControlSystemGroup
```
Should probably also be in the `EventsAndTriggersGroup`

knelli2 commented on src/ControlSystem/Event.hpp:
```diff
@@ -0,0 +1,125 @@
+    return domain::functions_of_time_are_ready<domain::Tags::FunctionsOfTime>(
+        cache, array_index, component, time);
+  }
+
+  bool needs_evolved_variables() const noexcept override { return true; }
```
LCOV

knelli2 commented on src/ControlSystem/Event.hpp:
```diff
@@ -0,0 +1,125 @@
+/// The list of events needed for measurements for a list of control
+/// systems.
+template <typename ControlSystems>
+using control_system_events = tmpl::transform<
+    metafunctions::measurements_t<ControlSystems>,
+    tmpl::bind<Event, metafunctions::control_systems_with_measurement<
+                          tmpl::pin<ControlSystems>, tmpl::_1>>>;
```
Should this be in the Metafunctions.hpp file?

wthrowe replied:
I looked into this, and it unfortunately creates some unpleasant include structure.  This file needs to include Metafunctions.hpp for stuff above.  Calling this metafunction technically only requires a forward declaration for the event, but using the result requires the full definition, so it should provide it, but that creates an include loop (unless we do a trailing include of Event.hpp in Metavariables.hpp, but those are ugly).

knelli2 commented on src/ControlSystem/RunCallbacks.hpp:
```diff
@@ -0,0 +1,54 @@
+
+namespace control_system {
+template <typename Submeasurement, typename ControlSystems>
+struct RunCallbacks {
```
Needs some docs on what this does exactly.

knelli2 commented on src/ControlSystem/RunCallbacks.hpp:
```diff
@@ -0,0 +1,54 @@
+      tt::assert_conforms_to<System, protocols::ControlSystem>>;
+  static_assert(tmpl::all<ControlSystems,
+                          tmpl::bind<assert_conforms_to_t, tmpl::_1>>::value);
+
+ public:
+  using observation_types = tmpl::list<>;
```
This isn't used anywhere. Is it intended to be a place holder for future revisions?

wthrowe replied:
It's used by the horizon finder if this is used as a callback there.

knelli2 commented on src/ControlSystem/RunCallbacks.hpp:
```diff
@@ -0,0 +1,54 @@
+                    Parallel::GlobalCache<Metavariables>& cache,
+                    const LinkedMessageId<double>& measurement_id) noexcept {
+    tmpl::for_each<ControlSystems>(
+        [&box, &cache, &measurement_id](auto control_system_v) noexcept {
+          using ControlSystem = tmpl::type_from<decltype(control_system_v)>;
+          db::apply<typename ControlSystem::process_measurement::
```
Comment about what component is this running on.

wthrowe replied:
Unspecified.  There's not much you could do with the information without access to the component box, anyway.

knelli2 commented on tests/Unit/Helpers/ControlSystem/Examples.hpp:
```diff
@@ -0,0 +1,137 @@
+    // the action testing framework doesn't support reductions, so
+    // this example is for a control system running on a singleton and
+    // we just pass the data through.
+    const auto box =
+        db::create<db::AddSimpleTags<MeasurementResultTag>>(data_from_element);
```
Not a fan of creating a temporary box here but not sure how else to do this.

knelli2 commented on tests/Unit/Helpers/ControlSystem/Examples.hpp:
```diff
@@ -0,0 +1,137 @@
+
+  // As with the previous struct, this is not part of the required
+  // interface.
+  struct MeasurementQueue : db::SimpleTag {
+    using type =
+        LinkedMessageQueue<double, tmpl::list<ExampleSubmeasurementQueueTag>>;
+  };
```
One thing I found useful was adding a type alias `simple_tags` to the control system so that the control component could easily access this queue to put in its databox. Also not required for the interface but may be useful for the example.

wthrowe replied:
If it's used by the component it should be part of the specified interface.  We can decide on that when the protocol is expanded for the component stuff.

knelli2 commented on tests/Unit/ControlSystem/Test_Measurement.cpp:
```diff
@@ -0,0 +1,112 @@
+  using metavariables = Metavariables;
+  using chare_type = ActionTesting::MockSingletonChare;
+  using array_index = int;
+  using initialization_tags = tmpl::list<
+      control_system::TestHelpers::ExampleControlSystem::MeasurementQueue,
+      Tags::Time, control_system::TestHelpers::MeasurementResultTag>;
```
For a test it probably doesn't matter, but I generally haven't needed to add the time to the control component databox. Also, I don't think MeasurementQueue is strictly needed for initialization (only needed in simple_tags), but I guess since you aren't running the SetupDataBox initialization action this is probably the best way to get tags into the databox.

wthrowe replied:
The time is used to check the test output, but it would probably be better to use a custom tag instead of the standard Time tag to make it clear that it isn't being used for anything else.

There are at least two seemingly equivalent methods of initializing DataBoxes in the tests, and I don't understand why or which one is preferred.

knelli2 reviewed (changes requested):
This is kind of a lot for a single commit, but I'm not sure of a good break point because everything depends on each other. The best I can come up with is putting the event in a separate commit from the protocols.

wthrowe commented on src/ControlSystem/Event.hpp:
```diff
@@ -0,0 +1,125 @@
+                const Component* const component) const noexcept {
+    // This checks all the functions of time, not just those the
+    // measurement is being made for.  We need all of them to access
+    // the evolved variables.
```
Change self-request: We need these to access the coordinates.  The variables themselves actually don't require them.

knelli2 commented on src/ControlSystem/Protocols/Submeasurement.hpp:
```diff
@@ -34,9 +34,9 @@ namespace control_system::protocols {
-/// the measurement must result in a single call on one chare to
-/// `control_system::RunCallbacks<ConformingStructBeingDefinedHere,
+/// function.  This function will be called on every element, and
+/// these calls must collectively result in a single call on one chare
```
Clarify that this "one chare" where the single call is can be any chare (not necessarily the elements or the control system chares). It just has to run *somewhere*. That confused me for a long time.

knelli2 reviewed (changes requested):
One more thing.

Also now that #3478 is merged, you'll have to fix some things. I got rid of the test helpers for the control system because there were only used for the updater, so you'll have to add those directories and CMakelists back in. Oops 😅

wthrowe commented:
Fixuped and rebased.  The `CMakeLists.txt` file in `tests/Unit/Helpers/ControlSystem` was largely rewritten in the rebase, but the others conflicts were trivial.

knelli2 commented on src/ControlSystem/Protocols/Submeasurement.hpp:
```diff
@@ -0,0 +1,65 @@
+/// where the last two are generally template parameters to the
+/// function.  This function will be called on every element, and
+/// these calls must collectively result in a single call on one chare
+/// (which need not be one of the element chares) to
+/// `control_system::RunCallbacks<ConformingStructBeingDefinedHere,
+/// ControlSystems>::apply`.  This will almost always require
```
Docs complain about this link to `control_system::RunCallbacks<ConformingStructBeingDefinedHere, ControlSystems>::apply`. I'm not sure how to fix this though

wthrowe replied:
I'm hoping someone else will have an idea.  It renders fine.

wthrowe replied:
And this time it passed?  Maybe something from rebasing on some of the recent doxygen changes.

knelli2 commented on tests/Unit/Helpers/ControlSystem/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+target_link_libraries(
+  ${LIBRARY}
+  INTERFACE
+  DataStructures
+  ControlSystem
```
Do we also keep libraries to link in alphabetical order?

knelli2 reviewed (commented):
You can squash these and the rest together.

kidder commented:
@nilsdeppe did you want to look at this PR?

nilsdeppe reviewed (approved):
I didn't go into everything in detail, mostly just read the docs. Looks good :)
